src/Engine/ProtoCore/Lang/JILFunctionEndPoint.cs

```diff
@@ -45,108 +45,114 @@ public override StackValue Execute(ProtoCore.Runtime.Context c, List<StackValue>
-            int locals = activation.locals;
-            
-
-            // Update the running block to tell the execution engine which set of instruction to execute
-            // TODO(Jun/Jiong): Considering store the orig block id to stack frame
-            int origRunningBlock = runtimeCore.RunningBlock;
-            runtimeCore.RunningBlock = svBlockDecl.BlockIndex;
-
-            StackFrameType callerType = stackFrame.CallerStackFrameType;
-
-            List<StackValue> registers = new List<DSASM.StackValue>();
-
-            StackValue svCallConvention;
-            bool isDispose = CoreUtils.IsDisposeMethod(procedureNode.Name);
-
-            bool explicitCall = !c.IsReplicating && !c.IsImplicitCall && !isDispose;
-            if (explicitCall)
-            {
-                svCallConvention = StackValue.BuildCallingConversion((int)ProtoCore.DSASM.CallingConvention.CallType.Explicit);
-            }
-            else
-            {
-                svCallConvention = StackValue.BuildCallingConversion((int)ProtoCore.DSASM.CallingConvention.CallType.Implicit);                
-            }
-
-            stackFrame.TX = svCallConvention;
-            interpreter.runtime.TX = svCallConvention;
-
-            // Set SX register 
-            stackFrame.BlockIndex = svBlockDecl;
-
-            // TODO Jun:
-            // The stackframe carries the current set of registers
-            // Determine if this can be done even for the non explicit call implementation
-            registers.AddRange(stackFrame.GetRegisters());
-
-
-            // Comment Jun: the depth is always 0 for a function call as we are reseting this for each function call
-            // This is only incremented for every language block bounce
-            int depth = stackFrame.Depth;
-            DSASM.StackFrameType type = stackFrame.StackFrameType;
-            Validity.Assert(depth == 0);
-            Validity.Assert(type == DSASM.StackFrameType.Function);
-
-            runtimeCore.RuntimeMemory.PushFrameForLocals(locals);
-            StackFrame newStackFrame = new StackFrame(svThisPtr, ci, fi, returnAddr, blockDecl, blockCaller, callerType, type, depth, framePointer, svBlockDecl.BlockIndex, registers, execStateSize);
-            runtimeCore.RuntimeMemory.PushStackFrame(newStackFrame);
-
-            StackValue svRet;
-
-            if (explicitCall)
-            {
-                svRet = ProtoCore.DSASM.StackValue.BuildExplicitCall(activation.pc);
-            }
-            else
-            {
-                svRet = interpreter.Run(runtimeCore.RunningBlock, activation.pc, Language.NotSpecified, runtimeCore.Breakpoints);
-                runtimeCore.RunningBlock = origRunningBlock;
-            }
-
-            if (runtimeCore.CurrentExecutive != null)
+            finally
             {
-                runtimeCore.CurrentExecutive.CurrentDSASMExec = oldDSASMExec;
+                if (runtimeCore.CurrentExecutive != null)
```

Comment:
okay, I think I get it -
when transpose thew an exception (or rather any function would throw an exception - I'm curious if transpose was special in some way? or the way it failed was special?)

then this switch back to the executive which I guess represents the original calling scope is never set back and chaos ensues.

Does this sound about right?

Reply:
You got it @mjkkirschner ! I guess this would have happened with any exception during the lapse the executive is replaced. Now with try/finally the executive should be switched back in any case.